File: src/linear_algebra_diagonalization_hard.py
"""
linear_algebra - diagonalization (hard)
Generated: 2026-02-22T06:00:31.678105
"""

from problem_utils import *

def generate():
    problem_type = choice(['diagonalizable_2x2', 'diagonalizable_3x3', 'eigenvector_find', 'power_matrix', 'not_diagonalizable'])
    
    if problem_type == 'diagonalizable_2x2':
        # Create a 2x2 matrix with known eigenvalues and eigenvectors
        lam1 = nonzero(-4, 4)
        lam2 = nonzero(-4, 4)
        while lam2 == lam1:
            lam2 = nonzero(-4, 4)
        
        # Create random eigenvector matrix P
        p11, p12 = nonzero(-3, 3), nonzero(-3, 3)
        p21, p22 = nonzero(-3, 3), nonzero(-3, 3)
        P = Matrix([[p11, p12], [p21, p22]])
        
        # Check if P is invertible
        if P.det() == 0:
            p22 = p22 + nonzero(1, 2)
            P = Matrix([[p11, p12], [p21, p22]])
        
        D = Matrix([[lam1, 0], [0, lam2]])
        A = P * D * P.inv()
        A = simplify(A)
        
        # Answer is the diagonal matrix
        ans = D
        
        return problem(
            question=f"Find the diagonal matrix $D$ such that $A = PDP^{{-1}}$ where $A = {latex(A)}$ and $P = {latex(P)}$",
            answer=fmt_matrix(ans),
            difficulty=(1600, 1750),
            topic="linear_algebra/diagonalization",
            answer_type="matrix",
            solution=steps(
                f"We need to find $D = P^{{-1}}AP$",
                f"First compute $P^{{-1}} = {latex(P.inv())}$",
                f"Then compute $P^{{-1}}A = {latex(P.inv() * A)}$",
                f"Finally $D = (P^{{-1}}A)P = {latex(ans)}$",
                f"The diagonal entries are the eigenvalues: ${lam1}$ and ${lam2}$"
            ),
            calculator="scientific"
        )
    
    elif problem_type == 'diagonalizable_3x3':
        # Create a 3x3 diagonalizable matrix with distinct eigenvalues
        lam1 = nonzero(-3, 3)
        lam2 = nonzero(-3, 3)
        lam3 = nonzero(-3, 3)
        while lam2 == lam1:
            lam2 = nonzero(-3, 3)
        while lam3 == lam1 or lam3 == lam2:
            lam3 = nonzero(-3, 3)
        
        # Simple eigenvectors
        P = Matrix([[1, 0, 1], [0, 1, 0], [1, 1, 1]])
        D = Matrix([[lam1, 0, 0], [0, lam2, 0], [0, 0, lam3]])
        A = P * D * P.inv()
        A = simplify(A)
        
        eigenvalues_set = FiniteSet(lam1, lam2, lam3)
        
        return problem(
            question=f"Find all eigenvalues of the matrix $A = {latex(A)}$",
            answer=fmt_set(eigenvalues_set),
            difficulty=(1650, 1800),
            topic="linear_algebra/diagonalization",
            answer_type="set",
            solution=steps(
                f"Compute the characteristic polynomial $\\det(A - \\lambda I)$",
                f"The characteristic equation is $(\\lambda - {lam1})(\\lambda - {lam2})(\\lambda - {lam3}) = 0$",
                f"The eigenvalues are $\\lambda_1 = {lam1}$, $\\lambda_2 = {lam2}$, $\\lambda_3 = {lam3}$"
            ),
            calculator="scientific"
        )
    
    elif problem_type == 'eigenvector_find':
        # Given matrix and eigenvalue, find eigenvector
        lam_known = nonzero(-3, 3)
        lam_other = nonzero(-3, 3)
        while lam_other == lam_known:
            lam_other = nonzero(-3, 3)
        
        # Construct matrix with known eigenvector [1, k]
        k = nonzero(-3, 3)
        v = Matrix([1, k])
        
        # A*v = lambda*v means [[a, b], [c, d]] * [[1], [k]] = lam_known * [[1], [k]]
        # a + b*k = lam_known
        # c + d*k = lam_known * k
        
        a_val = lam_known - k
        c_val = randint(-2, 2)
        d_val = lam_known * k - c_val
        
        A = Matrix([[a_val, 1], [c_val, d_val]])
        
        # Verify it's an eigenvector
        test = A * v - lam_known * v
        if test != Matrix([0, 0]):
            # Adjust to make it work
            A = Matrix([[lam_known, 0], [0, lam_known]]) + Matrix([[0, -k], [0, 0]])
        
        return problem(
            question=f"Find an eigenvector corresponding to eigenvalue $\\lambda = {lam_known}$ for the matrix $A = {latex(A)}$",
            answer=fmt_matrix(v),
            difficulty=(1600, 1700),
            topic="linear_algebra/diagonalization",
            answer_type="matrix",
            grading_mode="equivalent",
            solution=steps(
                f"Solve $(A - {lam_known}I)\\mathbf{{v}} = \\mathbf{{0}}$",
                f"$A - {lam_known}I = {latex(A - lam_known * eye(2))}$",
                f"Row reduce to find the null space",
                f"One eigenvector is $\\mathbf{{v}} = {latex(v)}$"
            ),
            calculator="scientific"
        )
    
    elif problem_type == 'power_matrix':
        # Use diagonalization to compute A^n
        lam1 = choice([2, -2, 3, -1])
        lam2 = choice([1, -1, 2])
        while lam2 == lam1:
            lam2 = choice([1, -1, 2])
        
        P = Matrix([[1, 1], [0, 1]])
        D = Matrix([[lam1, 0], [0, lam2]])
        A = P * D * P.inv()
        
        n_val = randint(5, 10)
        
        D_n = Matrix([[lam1**n_val, 0], [0, lam2**n_val]])
        A_n = P * D_n * P.inv()
        ans = simplify(A_n)
        
        return problem(
            question=f"If $A = {latex(A)} = PDP^{{-1}}$ where $P = {latex(P)}$ and $D = {latex(D)}$, compute $A^{{{n_val}}}$",
            answer=fmt_matrix(ans),
            difficulty=(1700, 1850),
            topic="linear_algebra/diagonalization",
            answer_type="matrix",
            solution=steps(
                f"Use the fact that $A^n = PD^nP^{{-1}}$",
                f"$D^{{{n_val}}} = {latex(D_n)}$",
                f"$A^{{{n_val}}} = P \\cdot D^{{{n_val}}} \\cdot P^{{-1}} = {latex(P)} {latex(D_n)} {latex(P.inv())}$",
                f"$A^{{{n_val}}} = {latex(ans)}$"
            ),
            calculator="scientific"
        )
    
    else:  # not_diagonalizable
        # Create a defective matrix (repeated eigenvalue without enough eigenvectors)
        lam = nonzero(-3, 3)
        # Jordan block form - not diagonalizable
        A = Matrix([[lam, 1], [0, lam]])
        
        ans = False
        
        return problem(
            question=f"Is the matrix $A = {latex(A)}$ diagonalizable over $\\mathbb{{R}}$?",
            answer=ans,
            difficulty=(1750, 1900),
            topic="linear_algebra/diagonalization",
            answer_type="boolean",
            solution=steps(
                f"Find eigenvalues by solving $\\det(A - \\lambda I) = 0$",
                f"$(\\lambda - {lam})^2 = 0$, so $\\lambda = {lam}$ with algebraic multiplicity 2",
                f"Find eigenvectors: $(A - {lam}I)\\mathbf{{v}} = {latex(A - lam*eye(2))}\\mathbf{{v}} = \\mathbf{{0}}$",
                f"The null space is 1-dimensional (geometric multiplicity 1)",
                f"Since geometric multiplicity $<$ algebraic multiplicity, $A$ is not diagonalizable"
            )
        )

emit(generate())

File: src/linear_algebra_diagonalization_hard.jl
# linear_algebra - diagonalization (hard)
# Generated: 2026-03-08T21:14:23.834183

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script begin
    set_topic!("linear_algebra/diagonalization")
    
    problem_type = choice(1:5)
    
    if problem_type == 1
        # Eigenvalue/eigenvector computation for 2x2 matrix (1800-2000)
        λ1 = randint(-8, 8)
        λ2 = randint(-8, 8)
        while λ2 == λ1
            λ2 = randint(-8, 8)
        end
        
        # Create eigenvectors
        v1 = [nonzero(-3, 3), nonzero(-3, 3)]
        v2 = [nonzero(-3, 3), nonzero(-3, 3)]
        while v1[1] * v2[2] == v1[2] * v2[1]  # ensure linearly independent
            v2 = [nonzero(-3, 3), nonzero(-3, 3)]
        end
        
        # Construct matrix A = PDP^(-1)
        P = [v1[1] v2[1]; v1[2] v2[2]]
        D = [λ1 0; 0 λ2]
        det_P = v1[1] * v2[2] - v1[2] * v2[1]
        P_inv = [v2[2] -v2[1]; -v1[2] v1[1]] .// det_P
        A = P * D * P_inv
        
        a11, a12 = Rational.(A[1, :])
        a21, a22 = Rational.(A[2, :])
        
        problem(
            question="Find all eigenvalues of the matrix \\(A = $(fmt_matrix([a11 a12; a21 a22]))\\). Enter your answer as a set.",
            answer=fmt_set(Set([λ1, λ2])),
            difficulty=(1800, 2000),
            answer_type="set",
            solution=steps(
                "Compute the characteristic polynomial \\(\\det(A - \\lambda I) = 0\\)",
                "\\(\\det\\begin{pmatrix}$(a11) - \\lambda & $(a12) \\\\ $(a21) & $(a22) - \\lambda\\end{pmatrix} = ($(a11) - \\lambda)($(a22) - \\lambda) - $(a12 * a21)\\)",
                "Expand and solve: \\(\\lambda^2 - $(a11 + a22)\\lambda + $(a11*a22 - a12*a21) = 0\\)",
                "Eigenvalues: \\(\\{$(λ1), $(λ2)\\}\\)"
            ),
            time=180
        )
        
    elseif problem_type == 2
        # Diagonalizability check (2000-2200)
        λ = randint(-6, 6)
        
        if choice([true, false])
            # Diagonalizable case: distinct eigenvalues or full set of eigenvectors
            λ2 = randint(-6, 6)
            while λ2 == λ
                λ2 = randint(-6, 6)
            end
            
            v1 = [nonzero(-3, 3), nonzero(-3, 3)]
            v2 = [nonzero(-3, 3), nonzero(-3, 3)]
            while v1[1] * v2[2] == v1[2] * v2[1]
                v2 = [nonzero(-3, 3), nonzero(-3, 3)]
            end
            
            P = [v1[1] v2[1]; v1[2] v2[2]]
            D = [λ 0; 0 λ2]
            det_P = v1[1] * v2[2] - v1[2] * v2[1]
            P_inv = [v2[2] -v2[1]; -v1[2] v1[1]] .// det_P
            A = P * D * P_inv
            
            is_diag = "Yes"
            explanation = "The matrix has two distinct eigenvalues (\\($(λ)\\) and \\($(λ2)\\)), so it has two linearly independent eigenvectors and is diagonalizable."
        else
            # Non-diagonalizable case: repeated eigenvalue with geometric multiplicity < algebraic multiplicity
            a = nonzero(-4, 4)
            A = [λ a; 0 λ]
            
            is_diag = "No"
            explanation = "The matrix has eigenvalue \\($(λ)\\) with algebraic multiplicity 2 but geometric multiplicity 1 (only one linearly independent eigenvector), so it is not diagonalizable."
        end
        
        a11, a12 = Rational.(A[1, :])
        a21, a22 = Rational.(A[2, :])
        
        problem(
            question="Is the matrix \\(A = $(fmt_matrix([a11 a12; a21 a22]))\\) diagonalizable? Answer Yes or No.",
            answer=is_diag,
            difficulty=(2000, 2200),
            answer_type="string",
            solution=steps(
                "Find the characteristic polynomial and eigenvalues",
                "Check the geometric multiplicity of each eigenvalue",
                explanation,
                "Answer: \\(\\text{$(is_diag)}\\)"
            ),
            time=240
        )
        
    elseif problem_type == 3
        # Matrix power using diagonalization (2100-2300)
        λ1 = choice([-2, -1, 2, 3])
        λ2 = choice([-2, -1, 2, 3])
        while λ2 == λ1
            λ2 = choice([-2, -1, 2, 3])
        end
        
        n = randint(5, 12)
        
        v1 = [1, nonzero(-2, 2)]
        v2 = [nonzero(-2, 2), 1]
        
        P = [v1[1] v2[1]; v1[2] v2[2]]
        D = [λ1 0; 0 λ2]
        det_P = v1[1] * v2[2] - v1[2] * v2[1]
        P_inv = [v2[2] -v2[1]; -v1[2] v1[1]] .// det_P
        A = P * D * P_inv
        
        # Compute A^n
        D_n = [λ1^n 0; 0 λ2^n]
        A_n = P * D_n * P_inv
        
        a11, a12 = Rational.(A[1, :])
        a21, a22 = Rational.(A[2, :])
        
        ans11, ans12 = Rational.(A_n[1, :])
        ans21, ans22 = Rational.(A_n[2, :])
        
        problem(
            question="Let \\(A = $(fmt_matrix([a11 a12; a21 a22]))\\). Compute \\(A^{$(n)}\\) and enter the element in row 1, column 1.",
            answer=ans11,
            difficulty=(2100, 2300),
            solution=steps(
                "Diagonalize \\(A = PDP^{-1}\\) where \\(D = $(fmt_matrix([λ1 0; 0 λ2]))\\)",
                "Then \\(A^{$(n)} = PD^{$(n)}P^{-1}\\)",
                "Compute \\(D^{$(n)} = $(fmt_matrix([λ1^n 0; 0 λ2^n]))\\)",
                "Multiply to get \\(A^{$(n)}_{11} = $(ans11)\\)"
            ),
            time=300
        )
        
    elseif problem_type == 4
        # Finding diagonal matrix D (1900-2100)
        λ1 = randint(-7, 7)
        λ2 = randint(-7, 7)
        while λ2 == λ1
            λ2 = randint(-7, 7)
        end
        
        v1 = [1, nonzero(-3, 3)]
        v2 = [nonzero(-3, 3), 1]
        
        P = [v1[1] v2[1]; v1[2] v2[2]]
        D = [λ1 0; 0 λ2]
        det_P = v1[1] * v2[2] - v1[2] * v2[1]
        P_inv = [v2[2] -v2[1]; -v1[2] v1[1]] .// det_P
        A = P * D * P_inv
        
        a11, a12 = Rational.(A[1, :])
        a21, a22 = Rational.(A[2, :])
        
        p11, p12 = Rational.(P[1, :])
        p21, p22 = Rational.(P[2, :])
        
        problem(
            question="Given \\(A = $(fmt_matrix([a11 a12; a21 a22]))\\) and \\(P = $(fmt_matrix([p11 p12; p21 p22]))\\), find the diagonal matrix \\(D\\) such that \\(A = PDP^{-1}\\). Enter the element \\(D_{11}\\) (top-left entry).",
            answer=λ1,
            difficulty=(1900, 2100),
            solution=steps(
                "We have \\(D = P^{-1}AP\\)",
                "The diagonal entries of \\(D\\) are the eigenvalues of \\(A\\)",
                "Since \\(P\\) has eigenvectors as columns, \\(D_{11}\\) corresponds to the eigenvalue for the first column of \\(P\\)",
                "Answer: \\(D_{11} = $(λ1)\\)"
            ),
            time=240
        )
        
    else  # problem_type == 5
        # Trace and determinant via eigenvalues (1800-2000)
        λ1 = randint(-8, 8)
        λ2 = randint(-8, 8)
        λ3 = randint(-8, 8)
        
        trace_val = λ1 + λ2 + λ3
        det_val = λ1 * λ2 * λ3
        
        quantity = choice(["trace", "determinant"])
        
        if quantity == "trace"
            ans = trace_val
            answer_str = "The trace equals the sum of eigenvalues: \\($(λ1) + $(λ2) + $(λ3) = $(trace_val)\\)"
        else
            ans = det_val
            answer_str = "The determinant equals the product of eigenvalues: \\(($(λ1))($(λ2))($(λ3)) = $(det_val)\\)"
        end
        
        problem(
            question="A \\(3 \\times 3\\) matrix \\(A\\) has eigenvalues \\($(λ1)\\), \\($(λ2)\\), and \\($(λ3)\\). Find the $(quantity) of \\(A\\).",
            answer=ans,
            difficulty=(1800, 2000),
            solution=steps(
                "Recall that for any matrix, the trace is the sum of eigenvalues and the determinant is the product of eigenvalues",
                answer_str,
                "Answer: \\($(ans)\\)"
            ),
            time=120
        )
    end
end